In [2]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn import preprocessing
from sklearn.metrics import mean_squared_error

from sklearn.preprocessing import PolynomialFeatures

from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge
from sklearn.ensemble import RandomForestRegressor
from sklearn.neural_network import MLPRegressor

In [3]:
data_path_file='DelayedFlights.csv'
df =pd.read_csv(data_path_file)

In [9]:
df.columns

Index(['Unnamed: 0', 'Year', 'Month', 'DayofMonth', 'DayOfWeek', 'DepTime',
       'CRSDepTime', 'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum',
       'TailNum', 'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
       'DepDelay', 'Origin', 'Dest', 'Distance', 'TaxiIn', 'TaxiOut',
       'Cancelled', 'CancellationCode', 'Diverted', 'CarrierDelay',
       'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'],
      dtype='object')

Centraremos el analisis en las caracteristicas de interés: features y lo guardaremos en el dataframe X

In [19]:
feat0=[ 'Month', 'DayofMonth', 'DayOfWeek', 'DepTime','CRSDepTime', 'CRSArrTime', 'CRSElapsedTime','DepDelay', 'ArrDelay',
       'Distance','TaxiIn', 'TaxiOut','CarrierDelay','WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay']

In [10]:
features=[ 'DayOfWeek', 'DepTime', 'CRSDepTime', 'ArrTime', 'CRSArrTime', 'ActualElapsedTime',
          'CRSElapsedTime', 'AirTime', 'ArrDelay','DepDelay', 'Distance', 'Cancelled', 'Diverted', 'UniqueCarrier']

In [11]:
X= df[features]

In [12]:
X.corr()

DayOfWeek   DepTime  CRSDepTime   ArrTime  CRSArrTime  \
DayOfWeek           1.000000  0.021924    0.027039  0.010913    0.017750   
DepTime             0.021924  1.000000    0.881598  0.458934    0.711513   
CRSDepTime          0.027039  0.881598    1.000000  0.396724    0.710303   
ArrTime             0.010913  0.458934    0.396724  1.000000    0.619385   
CRSArrTime          0.017750  0.711513    0.710303  0.619385    1.000000   
ActualElapsedTime   0.003072 -0.047040   -0.034925 -0.013595    0.033203   
CRSElapsedTime      0.004954 -0.044619   -0.026388 -0.012911    0.040117   
AirTime             0.004738 -0.054831   -0.036582 -0.017684    0.025907   
ArrDelay            0.006123  0.127017    0.044447 -0.050948    0.043078   
DepDelay            0.008538  0.139254    0.058875 -0.053024    0.053706   
Distance            0.008138 -0.056003   -0.029517 -0.027751    0.024335   
Cancelled           0.000023  0.002280    0.000023       NaN    0.000201   
Diverted           -0.001361 -0.004632   -0.009096 -0.007204    0.000907   

                   ActualElapsedTime  CRSElapsedTime   AirTime  ArrDelay  \
DayOfWeek                   0.003072        0.004954  0.004738  0.006123   
DepTime                    -0.047040       -0.044619 -0.054831  0.127017   
CRSDepTime                 -0.034925       -0.026388 -0.036582  0.044447   
ArrTime                    -0.013595       -0.012911 -0.017684 -0.050948   
CRSArrTime                  0.033203        0.040117  0.025907  0.043078   
ActualElapsedTime           1.000000        0.971122  0.976660  0.068130   
CRSElapsedTime              0.971122        1.000000  0.986086 -0.015676   
AirTime                     0.976660        0.986086  1.000000 -0.000189   
ArrDelay                    0.068130       -0.015676 -0.000189  1.000000   
DepDelay                    0.018492        0.007001 -0.002753  0.952927   
Distance                    0.952980        0.981759  0.980294 -0.029853   
Cancelled                        NaN       -0.003916       NaN       NaN   
Diverted                         NaN        0.024660       NaN       NaN   

                   DepDelay  Distance  Cancelled  Diverted  
DayOfWeek          0.008538  0.008138   0.000023 -0.001361  
DepTime            0.139254 -0.056003   0.002280 -0.004632  
CRSDepTime         0.058875 -0.029517   0.000023 -0.009096  
ArrTime           -0.053024 -0.027751        NaN -0.007204  
CRSArrTime         0.053706  0.024335   0.000201  0.000907  
ActualElapsedTime  0.018492  0.952980        NaN       NaN  
CRSElapsedTime     0.007001  0.981759  -0.003916  0.024660  
AirTime           -0.002753  0.980294        NaN       NaN  
ArrDelay           0.952927 -0.029853        NaN       NaN  
DepDelay           1.000000 -0.007718   0.016542  0.022867  
Distance          -0.007718  1.000000  -0.005060  0.021692  
Cancelled          0.016542 -0.005060   1.000000 -0.001146  
Diverted           0.022867  0.021692  -0.001146  1.000000

In [20]:
X0=df[feat0]

In [21]:
X0.shape

(1936758, 17)

In [22]:
X0.corr()

Month  DayofMonth  DayOfWeek   DepTime  CRSDepTime  \
Month              1.000000    0.059651   0.000088 -0.007809   -0.011367   
DayofMonth         0.059651    1.000000   0.017476  0.001014    0.001019   
DayOfWeek          0.000088    0.017476   1.000000  0.021924    0.027039   
DepTime           -0.007809    0.001014   0.021924  1.000000    0.881598   
CRSDepTime        -0.011367    0.001019   0.027039  0.881598    1.000000   
CRSArrTime         0.001367    0.000758   0.017750  0.711513    0.710303   
CRSElapsedTime     0.007046   -0.000028   0.004954 -0.044619   -0.026388   
DepDelay           0.004769    0.005289   0.008538  0.139254    0.058875   
ArrDelay          -0.000897    0.004129   0.006123  0.127017    0.044447   
Distance           0.005498    0.000117   0.008138 -0.056003   -0.029517   
TaxiIn             0.019705   -0.002926   0.008516 -0.012023   -0.032379   
TaxiOut            0.002574   -0.002201  -0.010302  0.030481    0.011408   
CarrierDelay       0.000420   -0.000947   0.010215 -0.051948   -0.107337   
WeatherDelay       0.006611    0.000916   0.005654  0.005307   -0.009338   
NASDelay           0.011441    0.005644  -0.006628  0.022530   -0.032701   
SecurityDelay     -0.003066   -0.000308   0.004665 -0.015273   -0.017064   
LateAircraftDelay  0.004953    0.009100   0.008839  0.169453    0.167734   

                   CRSArrTime  CRSElapsedTime  DepDelay  ArrDelay  Distance  \
Month                0.001367        0.007046  0.004769 -0.000897  0.005498   
DayofMonth           0.000758       -0.000028  0.005289  0.004129  0.000117   
DayOfWeek            0.017750        0.004954  0.008538  0.006123  0.008138   
DepTime              0.711513       -0.044619  0.139254  0.127017 -0.056003   
CRSDepTime           0.710303       -0.026388  0.058875  0.044447 -0.029517   
CRSArrTime           1.000000        0.040117  0.053706  0.043078  0.024335   
CRSElapsedTime       0.040117        1.000000  0.007001 -0.015676  0.981759   
DepDelay             0.053706        0.007001  1.000000  0.952927 -0.007718   
ArrDelay             0.043078       -0.015676  0.952927  1.000000 -0.029853   
Distance             0.024335        0.981759 -0.007718 -0.029853  1.000000   
TaxiIn               0.013819        0.099646  0.074748  0.156277  0.072284   
TaxiOut              0.037942        0.123839  0.078676  0.286484  0.070298   
CarrierDelay        -0.096156        0.026586  0.528038  0.501773  0.028443   
WeatherDelay        -0.005033       -0.016478  0.244488  0.264624 -0.023819   
NASDelay             0.009227        0.046964  0.233976  0.387606  0.018109   
SecurityDelay       -0.013349        0.006104 -0.002343 -0.002488  0.007029   
LateAircraftDelay    0.131467       -0.011584  0.513041  0.478123 -0.010578   

                     TaxiIn   TaxiOut  CarrierDelay  WeatherDelay  NASDelay  \
Month              0.019705  0.002574      0.000420      0.006611  0.011441   
DayofMonth        -0.002926 -0.002201     -0.000947      0.000916  0.005644   
DayOfWeek          0.008516 -0.010302      0.010215      0.005654 -0.006628   
DepTime           -0.012023  0.030481     -0.051948      0.005307  0.022530   
CRSDepTime        -0.032379  0.011408     -0.107337     -0.009338 -0.032701   
CRSArrTime         0.013819  0.037942     -0.096156     -0.005033  0.009227   
CRSElapsedTime     0.099646  0.123839      0.026586     -0.016478  0.046964   
DepDelay           0.074748  0.078676      0.528038      0.244488  0.233976   
ArrDelay           0.156277  0.286484      0.501773      0.264624  0.387606   
Distance           0.072284  0.070298      0.028443     -0.023819  0.018109   
TaxiIn             1.000000  0.066936     -0.021982      0.025661  0.229828   
TaxiOut            0.066936  1.000000     -0.028763      0.083243  0.427987   
CarrierDelay      -0.021982 -0.028763      1.000000     -0.068951 -0.119534   
WeatherDelay       0.025661  0.083243     -0.068951      1.000000 -0.002553   
NASDelay           0.229828  0.427987     -0.

In [13]:
X.shape

(1936758, 14)

Dado:

* la variablidad en los retrasos por compañias 
* el gran numero de datos disponibles
* Incluir la variable categorica 'UniqueCarrier' con 20 valores diferentes con un One-Hot Encoding seria ineficiente.

Escogeremos una compañia para realizar nuestro modelo.

Escogemos US Airways 'US' lo cual constituye aproximadamente el 5% de los vuelos.

In [6]:
X = X[ X['UniqueCarrier'] == 'US' ]

In [7]:
X.shape

(98425, 14)

Supongamos que **queremos entrenar un modelo que sea capaz de estimar la caracteristica 'ArrDelay' una vez el vuelo ya ha salido pero obviamente aun no ha llegado**. 

Para ello 'Arrdelay' no ha de tener faltantes. tratandose de un numero relativamente bajo descartaremos las filas con 'ArrDelay' faltantes para toda la matriz de caracteristicas X

In [8]:
nodelay = ~(X['ArrDelay'].isnull())
X= X[nodelay]

In [9]:
X.shape

(98007, 14)

Guardamos en la variable y la caracteristica 'ArrDelay' y la eliminaremos de la matriz de caracteristicas X

In [10]:
y= X.ArrDelay
X.drop( columns='ArrDelay', inplace= True )

In [11]:
X.shape

(98007, 13)

Por tanto las caracteristicas de la matrix X seran las contenidas en la lista feat_X. Por ser nulas Cancelled y Diverted se eliminan tambien. **Como el vuelo no ha llegado en el momento de estimar su retraso, no dispondremos de 'ArrTime', 'ActualElapsedTime' ni 'Airtime'**

In [12]:
feat_X=[ 'DayOfWeek', 'DepTime', 'CRSDepTime', 'CRSArrTime',
          'CRSElapsedTime','DepDelay', 'Distance']

In [13]:
X= X[feat_X]

In [14]:
X.shape

(98007, 7)

Para los modelos Lineales y RandomForest no es necesario standardizar los datos. No obstante **para el modelo MLPRegressor si es necesario para mejorar la convergencia.** Asi pues **standardizaremos los datos de la matriz X**

Como se standardiza en base a los datos de train como paso previo separamos con train_test_split()

In [15]:
# Dividimos los datos (X,y) en training y test (X_train, X_test, y_train, y_test)

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8, test_size=0.2, random_state=33)
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(78405, 7) (19602, 7) (78405,) (19602,)


Standardizamos ambos dataframes: X_train, X_test

In [16]:
# Instancia del scaler
scaler = preprocessing.StandardScaler()

In [17]:
# Fit to TRAIN data. Then transform it.
M_train = scaler.fit_transform(X_train)
#Volvemos a meter la matriz en dataframe
X_train = pd.DataFrame(M_train, columns= feat_X )

# Transform TEST data 
M_test = scaler.transform(X_test)
#Volvemos a meter la matriz en dataframe
X_test= pd.DataFrame(M_test, columns= feat_X )

In [18]:
print(X_train.shape, X_test.shape)

(78405, 7) (19602, 7)


--------------------------
## Ejercicio 1


Crea almenos 3 modelos de Regresion diferentes para intentar predecir el retraso de los vuelos 'ArrDelay'

-------------------------


Usaremos los siguientes modelos:

* Modelo Lineal
* Random Forest
* MLP Regressor

Generaremos instancias de los modelos, los entrenaremos y realizaremos las predicciones para todos ellos con los parametros por defecto.

In [19]:
# Modelo Lineal
linear_1 = LinearRegression()

#Entrenamos el modelo lineal
linear_1.fit(X_train, y_train)

#Predecimos a partir de los datos de test
y_pred_lin1= linear_1.predict(X_test)

In [20]:
# Modelo Random Forest
forest_1 = RandomForestRegressor()

#Entrenamos el Forest Regressor
forest_1.fit(X_train, y_train)

#Predecimos a partir de los datos de test
y_pred_for1 = forest_1.predict(X_test)

In [21]:
# Modelo MLPRegressor
mlp_1 = MLPRegressor()

#Entrenamos el MLP Regressor
mlp_1.fit(X_train, y_train)

#Predecimos a partir de los datos de test
y_pred_mlp1 = mlp_1.predict(X_test)

--------------------------
## Ejercicio 2

Compara los modelos segun el MSE y el R2.

--------------------------

Evaluaremos la calidad del modelo con los datos de test

In [22]:
r2_lin1 = linear_1.score(X_test, y_test)
mse_lin1 = mean_squared_error(y_test, y_pred_lin1)

r2_for1 = forest_1.score(X_test, y_test)
mse_for1 = mean_squared_error(y_test, y_pred_for1)

r2_mlp1= mlp_1.score(X_test, y_test)
mse_mlp1 = mean_squared_error(y_test, y_pred_mlp1)

In [23]:
print('Linear Regression | R^2: ', round(r2_lin1,6), ' | ', 'MSE: ', round(mse_lin1,6) )
print('Random Forest     | R^2: ', round(r2_for1,6), ' | ', 'MSE: ', round(mse_for1,6) )
print('MLPRegressor      | R^2: ', round(r2_mlp1,6), ' | ', 'MSE: ', round(mse_mlp1,6) )

Linear Regression | R^2:  0.863595  |  MSE:  354.007001
Random Forest     | R^2:  0.856217  |  MSE:  373.155913
MLPRegressor      | R^2:  0.866151  |  MSE:  347.374154


In [24]:
linear_1.coef_

array([-3.06781027e-02,  9.95221972e-01, -1.59531481e+00,  8.64666690e-01,
        1.13484463e+00,  4.63947048e+01, -3.17960071e+00])

El orden de los modelos de mejor a peor serian el MLPRegressor, Regresion Lineal y por ultimo Random Forest. Orden creciente de R^2 y decreciente de MSE.

---------------------------

## Ejercicio 3


Utiliza los diferentes parametros que admiten

----------------------------------------------

Para el modelo lineal implementaremos la variante Ridge que incluye en la funcion a minimizar la suma de los coeficientes al cuadrado.

Ridge Linear Regression:

* alpha: factor sobre la suma de coeficientes del modelo al cuadrado. alpha=0 seria el modelo lineal standard.

In [25]:
linear_2 = Ridge(alpha=1, random_state=33)

#Entrenamos el modelo lineal
linear_2.fit(X_train, y_train)

#Predecimos a partir de los datos de test
y_pred_lin2 = linear_2.predict(X_test)

In [26]:
# Modelo Random Forest
forest_2 = RandomForestRegressor(n_estimators = 200, 
                               max_depth=8, min_samples_split= 50, min_samples_leaf=20,
                               random_state = 33, criterion= 'squared_error', n_jobs=4,
                                verbose=0)

#Entrenamos el Forest Regressor
forest_2.fit(X_train, y_train)

#Predecimos a partir de los datos de test
y_pred_for2 = forest_2.predict(X_test)

In [69]:
# Modelo MLPRegressor
mlp_2 = MLPRegressor(hidden_layer_sizes= (140,20), activation='relu', solver='adam',
                      learning_rate_init= 0.005, learning_rate='constant', max_iter=200,
                      verbose=0, n_iter_no_change=10, tol=0.01, random_state=33)

#Entrenamos el MLP Regressor
mlp_2.fit(X_train, y_train)

#Predecimos a partir de los datos de test
y_pred_mlp2 = mlp_2.predict(X_test)

Iteration 1, loss = 330.27287621
Iteration 2, loss = 174.48419170
Iteration 3, loss = 173.77435812
Iteration 4, loss = 173.45426210
Iteration 5, loss = 172.89292245
Iteration 6, loss = 172.72522488
Iteration 7, loss = 172.82583959
Iteration 8, loss = 172.64346624
Iteration 9, loss = 172.96717558
Iteration 10, loss = 172.68139183
Iteration 11, loss = 172.08906230
Iteration 12, loss = 171.98396393
Iteration 13, loss = 171.89220632
Iteration 14, loss = 171.89937866
Iteration 15, loss = 171.66694029
Iteration 16, loss = 172.00945894
Iteration 17, loss = 171.34450158
Iteration 18, loss = 171.59752676
Iteration 19, loss = 171.50347744
Iteration 20, loss = 171.35942744
Iteration 21, loss = 171.10934088
Iteration 22, loss = 171.49879327
Iteration 23, loss = 170.94434586
Iteration 24, loss = 171.13240620
Iteration 25, loss = 171.58115997
Iteration 26, loss = 171.12105403
Iteration 27, loss = 170.84384335
Iteration 28, loss = 170.69853920
Iteration 29, loss = 170.61053422
Iteration 30, loss = 17

In [70]:
r2_lin2 = linear_2.score(X_test, y_test)
mse_lin2 = mean_squared_error(y_test, y_pred_lin2)

r2_for2 = forest_2.score(X_test, y_test)
mse_for2 = mean_squared_error(y_test, y_pred_for2)

r2_mlp2= mlp_2.score(X_test, y_test)
mse_mlp2 = mean_squared_error(y_test, y_pred_mlp2)

In [71]:
print('Linear Regression | R^2: ', round(r2_lin1,6), ' | ', 'MSE: ', round(mse_lin1,6) )
print('Ridge Linear Reg  | R^2: ', round(r2_lin2,6), ' | ', 'MSE: ', round(mse_lin2,6) )
print('Random Forest 1   | R^2: ', round(r2_for1,6), ' | ', 'MSE: ', round(mse_for1,6) )
print('Random Forest 2   | R^2: ', round(r2_for2,6), ' | ', 'MSE: ', round(mse_for2,6) )
print('MLPRegressor  1   | R^2: ', round(r2_mlp1,6), ' | ', 'MSE: ', round(mse_mlp1,6) )
print('MLPRegressor  2   | R^2: ', round(r2_mlp2,6), ' | ', 'MSE: ', round(mse_mlp2,6) )

Linear Regression | R^2:  0.863595  |  MSE:  354.007001
Ridge Linear Reg  | R^2:  0.863595  |  MSE:  354.006993
Random Forest 1   | R^2:  0.856217  |  MSE:  373.155913
Random Forest 2   | R^2:  0.864226  |  MSE:  352.368567
MLPRegressor  1   | R^2:  0.866151  |  MSE:  347.374154
MLPRegressor  2   | R^2:  0.866411  |  MSE:  346.698824


--------------------

## Ejercicio 4

Compara el rendimiento entre usar o no usar el train/test

----------------------

In [30]:
#Standardizamos todo el database
scaler2 = preprocessing.StandardScaler()
# Fit to ALL data. Then transform it.
M = scaler2.fit_transform(X)
#Volvemos a meter la matriz en el dataframe XX
XX = pd.DataFrame(M, columns= feat_X )

In [31]:
XX.shape

(98007, 7)

In [32]:
# Modelo Lineal
model1 = LinearRegression(n_jobs=4)
model1.fit(XX, y)
y_pred_1= model1.predict(XX)

# Modelo Random Forest
model2 = RandomForestRegressor()
model2.fit(XX, y)
y_pred_2= model2.predict(XX)

# Modelo MLPRegressor
model3 = MLPRegressor()
model3.fit(XX, y)
y_pred_3= model3.predict(XX)

c:\python38\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Evaluamos las predicciones

In [33]:
r2_m1= model1.score(XX, y)
mse1 = mean_squared_error(y, y_pred_1)
r2_m2= model2.score(XX, y)
mse2 = mean_squared_error(y, y_pred_2)
r2_m3= model3.score(XX, y)
mse3 = mean_squared_error(y, y_pred_3)

print('Linear Reg. All Data | R^2: ', round(r2_m1,6), '  | ', 'MSE:', round(mse1,6) )
print('Linear Reg.  Test D  | R^2: ', round(r2_lin1,6), ' | ', 'MSE: ', round(mse_lin1,6) )
print('Forest Reg. All Data | R^2: ', round(r2_m2,6), ' | ', 'MSE:', round(mse2,6) )
print('Forest Reg.  Test D  | R^2: ', round(r2_for1,6), ' | ', 'MSE: ', round(mse_for1,6) )
print('MLP Reg.    All Data | R^2: ', round(r2_m3,6), ' | ', 'MSE:', round(mse3,6) )
print('MLP Reg.    Test D   | R^2: ', round(r2_mlp1,6), ' | ', 'MSE: ', round(mse_mlp1,6) )

Linear Reg. All Data | R^2:  0.861602   |  MSE: 349.812151
Ridge Linear Test D  | R^2:  0.863595  |  MSE:  354.007001
Forest Reg. All Data | R^2:  0.977595  |  MSE: 56.629982
Random For2  Test D  | R^2:  0.856217  |  MSE:  373.155913
MLP Reg.  All Data   | R^2:  0.865594  |  MSE: 339.720798
MLPRegressor  2      | R^2:  0.866151  |  MSE:  347.374154


En general hacer la validacion con los mismos datos con los que se entrena el modelo dara R^2 mayor y MSE menor. Esto indicaria que las predicciones son mejores. No obstante es mejor hacer un test del modelo con datos que no hayan intervenido en la construccion del modelo. 

Asi se puede valorar la prediccion de datos que el modelo no ha usado. El sobre-entreno o 'Over Fitting' puede llevar a un modelo que prediga con precision los datos con los que fue construido pero le resulte dificil realizar una buena prediccion de datos que aun no ha visto.

-----------------------

## Ejercicio 5

Realiza algun proceso de Ingenieria de Variables para mejorar la prediccion.

-----------------------

Añadiremos la mejor variable predictora 'DepDelay' al cuadrado y al cubo en la matriz de caracteristicas. 

In [34]:
X_fe = X

In [35]:
feat_X_fe = feat_X
feat_X_fe.append('DD2')
feat_X_fe.append('DD3')
feat_X_fe

['DayOfWeek',
 'DepTime',
 'CRSDepTime',
 'CRSArrTime',
 'CRSElapsedTime',
 'DepDelay',
 'Distance',
 'DD2',
 'DD3']

In [36]:
X_fe['DD2']=X_fe['DepDelay']**2

In [37]:
X_fe['DD3']=X_fe['DepDelay']**3

In [38]:
X_fe.head()

DayOfWeek  DepTime  CRSDepTime  CRSArrTime  CRSElapsedTime  DepDelay  \
81557          2   1925.0        1855        2114           139.0      30.0   
81558          2    152.0         146         917           331.0       6.0   
81559          2   1407.0        1400        1514            74.0       7.0   
81560          2   2206.0        2142        2300           258.0      24.0   
81561          2   1453.0        1440        1703           143.0      13.0   

       Distance    DD2      DD3  
81557       646  900.0  27000.0  
81558      2551   36.0    216.0  
81559       227   49.0    343.0  
81560      1747  576.0  13824.0  
81561       665  169.0   2197.0

In [39]:
# Dividimos los datos (X,y) en training y test (X_train, X_test, y_train, y_test)

X_fe_train, X_fe_test, y_train, y_test = train_test_split(X_fe, y, train_size=0.8, test_size=0.2, random_state=33)
print(X_fe_train.shape, X_fe_test.shape, y_train.shape, y_test.shape)

(78405, 9) (19602, 9) (78405,) (19602,)


In [40]:
scaler3 = preprocessing.StandardScaler()

In [41]:
# Fit to TRAIN data. Then transform it.
M_fe_train = scaler3.fit_transform(X_fe_train)
#Volvemos a meter la matriz en dataframe
X_fe_train = pd.DataFrame(M_fe_train, columns= feat_X_fe )

# Transform TEST data 
M_fe_test = scaler3.transform(X_fe_test)
#Volvemos a meter la matriz en dataframe
X_fe_test= pd.DataFrame(M_fe_test, columns= feat_X_fe )

In [42]:
# Modelo Lineal
model1b = LinearRegression(n_jobs=4)

#Entrenamos el modelo lineal
model1b.fit(X_fe_train, y_train)

y_pred_1b= model1b.predict(X_fe_test)

In [43]:
r2_m1b= model1b.score(X_fe_test, y_test)
mse1b = mean_squared_error(y_test, y_pred_1b)

print('Linear Regression X_fe Validation Results: R^2:', round(r2_m1b,6), 'MSE:', round(mse1b,6) )

Linear Regression X_fe Validation Results: R^2: 0.863624 MSE: 353.930786


Observamos como la introduccion de 'DepDelay' al cuadrado y al cubo ha mejorado levemente la prediccion del modelo. El modelo lineal sin estas variables habia dado R^2:0.863595  y MSE:354.007001 Observamos un valor de R^2 mayor y menor MSE para las predicciones realizadas sobre el conjunto de datos de test.



----------------------------

## Ejercicio 6

No utilices la variable 'DepDelay' en las prediciones

----------------------------

Definamos la lista de variables a usar X_train_red y X_test_red 

In [44]:
X_train_red = X_train.drop( columns='DepDelay', inplace= False )
X_test_red = X_test.drop(columns='DepDelay', inplace=False )

In [45]:
# Modelo Lineal
model1c = LinearRegression(n_jobs=4)

# Modelo Random Forest
model2c = RandomForestRegressor(n_estimators = 200, 
                               max_depth=10, min_samples_split= 50, min_samples_leaf=20,
                               random_state = 33, criterion= 'squared_error', n_jobs=4)

# Modelo MLPRegressor
model3c = MLPRegressor(hidden_layer_sizes= (20,10,), activation='relu', solver='adam',
                      learning_rate_init= 0.05, learning_rate='adaptive', max_iter=100,
                      random_state=33)

In [46]:
#Entrenamos el modelo lineal
model1c.fit(X_train_red, y_train)


LinearRegression(n_jobs=4)

In [47]:
y_pred_1c= model1c.predict(X_test_red)
r2_m1c = model1c.score(X_test_red, y_test)
mse1c = mean_squared_error(y_test, y_pred_1c)

print('Linear Regression noDepDelay Validation Results: R^2:', round(r2_m1c,6), 'MSE:', round(mse1c,6) )

Linear Regression noDepDelay Validation Results: R^2: 0.017154 MSE: 2550.745076


Claramente obtenemos peor resultado en las predicciones. Menor R^2 y mayor MSE